### PR TITLE
crimson/osd: accurately judge whether to do drop_recovery_read on recovering objs

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Fixes: https://tracker.ceph.com/issues/47313
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by xxhdx1985126** [comment]
@tchaikov Please take a look:-)

File: src/crimson/osd/recovery_backend.h [hunk 1]
```diff
@@ -126,6 +126,7 @@ class RecoveryBackend {
 
     crimson::osd::ObjectContextRef obc;
     PullInfo pi;
```
Comment (tchaikov) [inline src/crimson/osd/recovery_backend.h:128]: what about making `pi` an `optional<PullInfo>` or `unique_ptr<PullInfo>` if it's not always valid?

File: src/crimson/osd/replicated_recovery_backend.cc [hunk 2]
```diff
@@ -32,7 +32,11 @@ seastar::future<> ReplicatedRecoveryBackend::recover_object(
     return maybe_pull_missing_obj(soid, need).then([this, soid](bool pulled) {
       return load_obc_for_recovery(soid, pulled);
     }).then([this, soid, need, &pops, &shards] {
-      return prep_push(soid, need, &pops, shards);
+      if (!sharrds.empty()) {
```
Comment (tchaikov) [inline src/crimson/osd/replicated_recovery_backend.cc:35]: typo.

File: src/crimson/osd/recovery_backend.h (hunk 1 above)
Comment (xxhdx1985126) [reply src/crimson/osd/recovery_backend.h:128]: Oh, yes!

**Comment by xxhdx1985126** [comment]
@tchaikov Please take a look again:)

**Review by tchaikov — APPROVED** [review]

**Pull request merged**